bpo-37702: Fix SSL certificate-store-handles leak

neonene opened the pull request:
In Windows, CertCloseStore(hCollectionStore,[any option]) closes only hCollectionStore.
To avoid out-of-memory, CertCloseStore() shuld be called to each hSystemStore
which was added to hCollectionStore.

I think CERT_CLOSE_STORE_FORCE_FLAG is rare option and not much needed.
MSDN says CertCloseStore(handle, 0) is ideal and works fine for me.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-37702](https://bugs.python.org/issue37702) -->
https://bugs.python.org/issue37702
<!-- /issue-number -->

neonene committed:
bpo-37702: Fix SSL certificate-store-handles leak

In Windows, CertCloseStore(hCollectionStore) closes only hCollectionStore.
To avoid out-of-memory, CertCloseStore() shuld be called to each hSystemStore
which was added to hCollectionStore.

epicfaace commented on Modules/_ssl.c:
```diff
@@ -5767,10 +5801,20 @@ _ssl_enum_crls_impl(PyObject *module, const char *store_name)
+       with CertCloseStore using CERT_CLOSE_STORE_FORCE_FLAG,
+       the collection store must be closed before its sibling stores.
+      (https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certaddstoretocollection) */
+    success = CertCloseStore(hCollectionStore, CERT_CLOSE_STORE_FORCE_FLAG);
```
You could remove the `BOOL success;` declaration before and do this?:
```suggestion
    BOOL success = CertCloseStore(hCollectionStore, CERT_CLOSE_STORE_FORCE_FLAG);
```

epicfaace commented on Misc/NEWS.d/next/Windows/2019-07-29-16-49-31.bpo-37702.Lj2f5e.rst:
```diff
@@ -0,0 +1 @@
+Fix memory leak in ssl certification.
```
It might be good to specify which methods this applies to, and that it is only on Windows.

neonene replied:
I'd rather not describe c-implementation level information.

neonene committed:
Update 2019-07-29-16-49-31.bpo-37702.Lj2f5e.rst

neonene committed:
Update Modules/_ssl.c

Co-Authored-By: Ashwin Ramaswami <aramaswamis@gmail.com>

neonene committed:
Update _ssl.c

epicfaace commented on Modules/_ssl.c:
```diff
@@ -5767,10 +5799,20 @@ _ssl_enum_crls_impl(PyObject *module, const char *store_name)
-       associated with the store, in this case our collection store and the
-       associated system stores. */
-    if (!CertCloseStore(hCollectionStore, CERT_CLOSE_STORE_FORCE_FLAG)) {
+    /* When a collection store and its sibling stores are closed
```
Do you think it would be good to keep in the definition of CERT_CLOSE_STORE_FORCE_FLAG ("forces freeing of memory for all contexts associated with the store")?

neonene replied:
Thanks for Bool position.
For now, both 0 and CERT_CLOSE_STORE_FORCE_FLAG  work fine for my pc.
I'd like to keep changes to a minimum and leave it to other ssl experts' judgement whether the flag should revert to 0 or not.

epicfaace replied:
Oh, I didn't mean reverting the flag to 0, I meant, did you want to keep the part in the comment that says "CERT_CLOSE_STORE_FORCE_FLAG forces freeing of memory for all contexts associated with the store"?

neonene replied:
MS says,
CERT_CLOSE_STORE_FORCE_FLAG:
Forces the freeing of memory for all contexts associated with the store.
This flag can be safely used only when the store is opened in a function
and neither the store handle nor any of its contexts are passed to
any called functions. For details, see ...  

I think, keeping short sentences caused and will cause a misunderstanding that
this rarely used flag is always safe and releases all certificate-store-handles.
Having said that, I'm not in favor of long description in the source file.

neonene committed:
Update _ssl.c

MS  doesn't recommend using CERT_CLOSE_STORE_FORCE_FLAG.
https://blogs.msdn.microsoft.com/winsdk/2010/01/29/passing-the-flag-cert_close_store_force_flag-to-certclosestore-may-cause-your-application-to-crash/

And I don't want to leave confusing comments on the flag that is no longer necessary to close all handles.
So, I decided to change the flag back to 0.

neonene committed:
bpo-37702: comment for function call to each handle

neonene committed:
close handles in ssl_collect_certificates()

neonene committed:
comments (correct english)

neonene committed:
indent, comments (correct english)

tiran commented on Modules/_ssl.c:
```diff
@@ -5626,12 +5624,25 @@ _ssl_enum_certificates_impl(PyObject *module, const char *store_name)
     PCCERT_CONTEXT pCertCtx = NULL;
     PyObject *keyusage = NULL, *cert = NULL, *enc = NULL, *tup = NULL;
     PyObject *result = NULL;
+    static DWORD system_stores[] = {
```
I don't like code duplication. Please make the system store list a global static

tiran commented on Modules/_ssl.c:
```diff
@@ -5626,12 +5624,25 @@ _ssl_enum_certificates_impl(PyObject *module, const char *store_name)
+        CERT_SYSTEM_STORE_CURRENT_USER_GROUP_POLICY,
+        CERT_SYSTEM_STORE_SERVICES,
+        CERT_SYSTEM_STORE_USERS};
+    size_t i, system_stores_count = sizeof(system_stores) / sizeof(DWORD);
```
This style is confusing and hard to read. Split ``size_t i`` and ``size_t system_stores_count`` into two lines.

tiran commented on Modules/_ssl.c:
```diff
@@ -5686,15 +5697,21 @@ _ssl_enum_certificates_impl(PyObject *module, const char *store_name)
-       associated system stores. */
-    if (!CertCloseStore(hCollectionStore, CERT_CLOSE_STORE_FORCE_FLAG)) {
+    /* CertCloseStore() with any argument can't close handles together.
+       In case of using CERT_CLOSE_STORE_FORCE_FLAG (not recommended),
```
The comment is out of date. You are referencing ``CERT_CLOSE_STORE_FORCE_FLAG`` but no longer using ``CERT_CLOSE_STORE_FORCE_FLAG`` as flag.

neonene committed:
make the system store list a global static

neonene committed:
surround #define replacement with parenthesis

Not necessary for now.

neonene committed:
add a function to close handles
